### PR TITLE
Enable WRSQ for locality and WRR LB

### PR DESCRIPTION
**Pull request by tonya11en**
### Enable the WRSQ scheduler for the locality and WRR LB

The WRSQ scheduler has replaced EDF for the locality and WRR load balancers. This fixes the problems outlined in https://github.com/envoyproxy/envoy/issues/14597 and introduces non-deterministic weighted random selection behavior.

WRSQ is now used by default and can be disabled by setting the following false: 
`envoy.reloadable_features.upstream.round_robin_scheduler_wrsq`
`envoy.reloadable_features.upstream.locality_scheduler_wrsq`

**Risk Level:**
Medium/high. The WRR and locality load balancers experience a change in their selection behavior such that it's non-deterministic.

**Testing:**
Existing unit tests. However, it's worth noting that tests are not run twice to cover behavior with the runtime guards toggled to use EDF for WRR and locality.

**Docs Changes:**
Done.

**Release Notes:**
Done.

**Runtime guard:**
WRSQ is used by default. This can be disabled with `envoy.reloadable_features.upstream.round_robin_scheduler_wrsq` and `envoy.reloadable_features.upstream.locality_scheduler_wrsq` being set to false.

Fixes https://github.com/envoyproxy/envoy/issues/14597

### CODE REVIEW

File: source/common/upstream/wrsq_scheduler.h [hunk 1]
```diff
@@ -114,7 +122,11 @@ class WRSQScheduler : public Scheduler<C>, protected Logger::Loggable<Logger::Id
     maybeRebuildCumulativeWeights();
 
     const double weight_sum = cumulative_weights_.back().cumulative_weight;
-    uint64_t rnum = random_.random() % static_cast<uint32_t>(weight_sum);
+
+    // It's possible that the weight sum can be fractional, so we need a random double.
+    const double r = static_cast<double>(random_.random() % accuracy()) / accuracy();
+    const double rnum = r * weight_sum;
```
Comment (antoniovicente) [inline source/common/upstream/wrsq_scheduler.h:128]: I think this algorithm breaks down in cases weight_sum > accuracy() / N.  I'm not sure how big weight_sum would need to be for problems to be noticiable.  weight_sum > accuracy() would result some elements being impossible to represent.

weight_sum does involve a sum of weight_val * it.second.size(), so I would expect weight_sum to be relatively large in cases where there is a large number of hosts.

Consider that the original algorithm that does mod weight_sum will produce exactly the same result as the new algorithm in cases where the test random source returns consecutive integers starting at 0.

File: test/common/upstream/load_balancer_impl_test.cc [hunk 2]
```diff
@@ -898,40 +899,50 @@ TEST_P(RoundRobinLoadBalancerTest, Weighted) {
...
-  EXPECT_EQ(hostSet().healthy_hosts_[0], lb_->chooseHost(nullptr));
-  EXPECT_EQ(hostSet().healthy_hosts_[1], lb_->chooseHost(nullptr));
-  EXPECT_EQ(hostSet().healthy_hosts_[1], lb_->chooseHost(nullptr));
-  EXPECT_EQ(hostSet().healthy_hosts_[0], lb_->chooseHost(nullptr));
-  EXPECT_EQ(hostSet().healthy_hosts_[1], lb_->chooseHost(nullptr));
+  hostSet().runCallbacks({hostSet().healthy_hosts_.back()}, {});
+  run_selections(runs);
+  EXPECT_NEAR(counts[0] / runs, 1.0 / 3, tolerance);
+  EXPECT_NEAR(counts[1] / runs, 2.0 / 3, tolerance);
+
```
Comment (antoniovicente) [inline test/common/upstream/load_balancer_impl_test.cc:928]: EXPECT_EQ(counts[2], 0);

File: test/common/upstream/load_balancer_impl_test.cc [hunk 3]
```diff
@@ -898,40 +899,50 @@ TEST_P(RoundRobinLoadBalancerTest, Weighted) {
...
   // Initial weights respected.
-  EXPECT_EQ(hostSet().healthy_hosts_[1], lb_->chooseHost(nullptr));
-  EXPECT_EQ(hostSet().healthy_hosts_[0], lb_->chooseHost(nullptr));
-  EXPECT_EQ(hostSet().healthy_hosts_[1], lb_->chooseHost(nullptr));
-  EXPECT_EQ(hostSet().healthy_hosts_[1], lb_->chooseHost(nullptr));
-  EXPECT_EQ(hostSet().healthy_hosts_[0], lb_->chooseHost(nullptr));
-  EXPECT_EQ(hostSet().healthy_hosts_[1], lb_->chooseHost(nullptr));
+  hostSet().runCallbacks({hostSet().healthy_hosts_.back()}, {});
+  run_selections(runs);
+  EXPECT_NEAR(counts[0] / runs, 1.0 / 3, tolerance);
```
Comment (antoniovicente) [inline test/common/upstream/load_balancer_impl_test.cc:926]: Worth writing instead:
EXPECT_NEAR(counts[0], 1.0 * runs / 3, runs * tolerance);
EXPECT_NEAR(counts[1], 2.0 * runs / 3, runs * tolerance);

File: test/common/upstream/subset_lb_test.cc [hunk 4]
```diff
@@ -1896,21 +1896,33 @@ TEST_F(SubsetLoadBalancerTest, EnabledLocalityWeightAwareness) {
...
+    EXPECT_CALL(random_, random()).WillRepeatedly(Return(i));
+    auto h = lb_->chooseHost(&context);
+    auto& hosts0 = host_set_.healthy_hosts_per_locality_->get()[0];
+    auto& hosts1 = host_set_.healthy_hosts_per_locality_->get()[1];
+    if (std::count(hosts0.begin(), hosts0.end(), h) == 1) {
+      counts[0]++;
+    } else if (std::count(hosts1.begin(), hosts1.end(), h) == 1) {
+      counts[1]++;
+    }
+  }
```
Comment (antoniovicente) [inline test/common/upstream/subset_lb_test.cc:1920]: the loop above seems to appear 3 times in this file.  Worth adding a helper method?

File: test/common/upstream/subset_lb_test.cc [hunk 5]
```diff
@@ -1896,21 +1896,33 @@ TEST_F(SubsetLoadBalancerTest, EnabledLocalityWeightAwareness) {
...
   TestLoadBalancerContext context({{"version", "1.1"}});
 
-  // Since we respect locality weights, the second locality is selected.
-  EXPECT_CALL(random_, random()).WillOnce(Return(0));
-  EXPECT_EQ(host_set_.healthy_hosts_per_locality_->get()[1][0], lb_->chooseHost(&context));
+  // Verify we respect locality weights.
+  const size_t iters = WRSQScheduler<int>::accuracy();
+  std::vector<size_t> counts({0, 0});
+  for (size_t i = 0; i < iters; ++i) {
+    EXPECT_CALL(random_, random()).WillRepeatedly(Return(i));
```
Comment (antoniovicente) [inline test/common/upstream/subset_lb_test.cc:1911]: How many times do we expect random() to be called per iteration?  Only once?

File: test/common/upstream/subset_lb_test.cc [hunk 6]
```diff
@@ -1942,15 +1954,19 @@ TEST_F(SubsetLoadBalancerTest, EnabledScaleLocalityWeights) {
...
+    auto h = lb_->chooseHost(&context);
+    if (h == host_set_.healthy_hosts_per_locality_->get()[0][1]) {
+      counts[0]++;
+    } else if (h == host_set_.healthy_hosts_per_locality_->get()[1][3]) {
+      counts[1]++;
+    }
+  }
+  EXPECT_NEAR(2.0 * iters / 3, counts[0], 1);
+  EXPECT_NEAR(iters / 3.0, counts[1], 1);
```
Comment (antoniovicente) [inline test/common/upstream/subset_lb_test.cc:1969]: The determinism in this loop comes from random() returning the positive integers in order.  You could do an exact expectation.  When iters is a multiple of 3:
EXPECT_EQ(2 * iters / 3, counts[0]);
EXPECT_EQ(iters / 3, counts[1]);

**Review by antoniovicente — COMMENTED** [review]
Excellent change.  I added some questions about the implementation specifics in order to fill gaps in my understanding, and some suggestions on how to add randomness in order to further reduce sync after rebuilding the host list.

File: test/common/upstream/subset_lb_test.cc [hunk 7]
```diff
@@ -1984,13 +2000,23 @@ TEST_F(SubsetLoadBalancerTest, EnabledScaleLocalityWeightsRounding) {
...
   // We expect to see a 33/66 split because 2 * 1 / 2 = 1 and 2 * 3 / 4 = 1.5 -> 2
-  EXPECT_EQ(host_set_.healthy_hosts_per_locality_->get()[1][0], lb_->chooseHost(&context));
-  EXPECT_EQ(host_set_.healthy_hosts_per_locality_->get()[0][0], lb_->chooseHost(&context));
-  EXPECT_EQ(host_set_.healthy_hosts_per_locality_->get()[1][1], lb_->chooseHost(&context));
-  EXPECT_EQ(host_set_.healthy_hosts_per_locality_->get()[1][2], lb_->chooseHost(&context));
-  EXPECT_EQ(host_set_.healthy_hosts_per_locality_->get()[0][0], lb_->chooseHost(&context));
-  EXPECT_EQ(host_set_.healthy_hosts_per_locality_->get()[1][0], lb_->chooseHost(&context));
-  EXPECT_EQ(host_set_.healthy_hosts_per_locality_->get()[1][1], lb_->chooseHost(&context));
+  std::vector<size_t> counts({0, 0});
+  const size_t iters = WRSQScheduler<int>::accuracy();
```
Comment (antoniovicente) [inline test/common/upstream/subset_lb_test.cc:2004]: I think that the test below will fail if you change this line to:
`const size_t iters = 10 * WRSQScheduler<int>::accuracy();`

The reason behind this is that with sum_weights = 3 and accuracy of 10000, you accumulate 1 error every 10,000 random inputs processed.  The original mod-based random implementation doesn't accumuluate errors over the accuracy period.

**Comment by jmarantz** [comment]
Looks like this needs a 'main merge' and there are a few outstanding comments. @tonya11en are you still working on this?

Thanks!

**Comment by tonya11en** [comment]
> Looks like this needs a 'main merge' and there are a few outstanding comments. @tonya11en are you still working on this?

I was waiting for the slow start patch to merge, since it was going to cause a whole bunch of merge conflicts. I'm still planning to get back to this.

File: source/common/upstream/wrsq_scheduler.h (hunk 1 above)
Comment (tonya11en) [reply source/common/upstream/wrsq_scheduler.h:128]: Oh wow, that's a good catch. I'll need to think about the best way to go about this..

**Review by tonya11en — COMMENTED** [review]
Did some rebasing and addressed a few comments. Still have some work to do, so stand by for more commits.

/wait

**Pull request closed without merging**